### ISSUE TITLE
Golang filter memory consumption: http.requestMap constantly grows

### ISSUE THREAD

**Issue by eerakli** [envoyproxy/envoy#36891]
*Title*: *Golang filter memory consumption: http.requestMap constantly grows*

*Description*:
I implemented golang filter for envoy, and during load test I found out that envoy-proxy memory consumption constantly grows.
During profiling I figured out that the reason is in the `var Requests = requestMap` that is located in https://github.com/envoyproxy/envoy/blob/v1.27.5/contrib/golang/filters/http/source/go/pkg/http/shim.go 
So library adds request to this map in `createRequest` method and removes request from this map in `envoyGoFilterOnHttpDestroy` method.
I've added logs in `StreamFilter.OnDestroy` implementation and looks like this method is not invoked even if request was processed.

Please, could you clarify what triggers `envoyGoFilterOnHttpDestroy` method? As I understand that's the only way to clear `requestMap`.

Envoy version is v1.27.5 (due to istio version limitations).

Filter's logic is located mostly in `DecodeHeaders` method: if everything is ok method returns `api.Continue` and requests goes to the next filter, or it responds with `api.LocalReply` with 403 status.

**Comment by spacewander** [comment]
envoyGoFilterOnHttpDestroy is expected to be called once the request has hit the Golang filter. Would you provide a standalone reproducible example so we are able to reproduce it?

There is a memory fix in Envoy 1.28: https://github.com/envoyproxy/envoy/commit/0a2d0f7e89ca53c30fd42617c667160c62cfa400, though it didn't touch the requestMap.

**Comment by eerakli** [comment]
Here's filter's source code. Also I have full setup in archive to reproduce this problem.
```go
package main

import (
	"fmt"
	"net/http"
	"os"
	"runtime"
	"runtime/pprof"
	"strings"
	"time"

	"github.com/envoyproxy/envoy/contrib/golang/common/go/api"
	envoyHttp "github.com/envoyproxy/envoy/contrib/golang/filters/http/source/go/pkg/http"
	"google.golang.org/protobuf/types/known/anypb"
)

const (
	// pluginName should be the same as in envoy.yaml configuration.
	pluginName = "auth_filter"
)

func init() {
	go CollectProfile()
	// register filter factory.
	envoyHttp.RegisterHttpFilterConfigFactoryAndParser(pluginName, NewFilterFactory(), NullParser{})
}

func main() {}

// NewFilterFactory returns a factory for the filter.
func NewFilterFactory() api.StreamFilterConfigFactory {
	return func(config interface{}) api.StreamFilterFactory {
		return func(callbacks api.FilterCallbackHandler) api.StreamFilter {
			return New(callbacks)
		}
	}
}

// NewPassThroughFactory returns a factory for the pass-through filter. If application fails to initialize the filter,
// it will use the pass-through filter to avoid extra logs in envoy that registered filter is not found.
func NewPassThroughFactory(reason string) api.StreamFilterConfigFactory {
	return func(config interface{}) api.StreamFilterFactory {
		return func(callbacks api.FilterCallbackHandler) api.StreamFilter {
			callbacks.Log(api.Warn, fmt.Sprintf("auth filter initialization error: %v", reason))
			return &api.PassThroughStreamFilter{}
		}
	}
}

// AuthorizationFilter defines istio behavior: it parses headers and returns 403/401 if user is banned or token is invalid.
type AuthorizationFilter struct {
	callbacks api.FilterCallbackHandler
	token     string
}

// New creates instance of AuthorizationFilter.
func New(callbacks api.FilterCallbackHandler) AuthorizationFilter {
	return AuthorizationFilter{
		callbacks: callbacks,
	}
}

func (f AuthorizationFilter) DecodeHeaders(header api.RequestHeaderMap, endStream bool) api.StatusType {
	rawToken, hasAuthorization := header.Get("authorization")
	if !hasAuthorization {
		f.callbacks.Log(api.Error, "no auth header")
		f.callbacks.SendLocalReply(http.StatusUnauthorized, "no auth header", nil, -1, "")
		return api.LocalReply
	}
	parts := strings.Split(rawToken, " ")
	if len(parts) < 2 {
		f.callbacks.Log(api.Error, "invalid header")
		f.callbacks.SendLocalReply(http.StatusUnauthorized, "invalid header", nil, -1, "")
		return api.LocalReply
	}
	jwtToken := parts[1]
	if jwtToken == "" {
		f.callbacks.Log(api.Error, "empty token")
		f.callbacks.SendLocalReply(http.StatusUnauthorized, "empty token", nil, -1, "empty token")
		return api.LocalReply
	}
	f.token = jwtToken
	f.callbacks.Log(api.Info, fmt.Sprintf("received token: %s", jwtToken))
	return api.Continue
}

func (f AuthorizationFilter) DecodeData(buffer api.BufferInstance, endStream bool) api.StatusType {
	return api.Continue
}

func (f AuthorizationFilter) DecodeTrailers(trailers api.RequestTrailerMap) api.StatusType {
	return api.Continue
}

func (f AuthorizationFilter) EncodeHeaders(header api.ResponseHeaderMap, endStream bool) api.StatusType {
	if f.token != "" {
		header.Set("X-Token", f.token)
	}
	return api.Continue
}

func (f AuthorizationFilter) EncodeData(buffer api.BufferInstance, endStream bool) api.StatusType {
	return api.Continue
}

func (f AuthorizationFilter) EncodeTrailers(trailers api.ResponseTrailerMap) api.StatusType {
	return api.Continue
}

func (f AuthorizationFilter) OnDestroy(reason api.DestroyReason) {
	info := f.callbacks.StreamInfo()
	rsn := "normal"
	if reason == api.Terminate {
		rsn = "terminate"
	}
	msg := fmt.Sprintf(
		"on destroy reason: `%s`; route name: `%s`; filter chain name: `%s`",
		rsn, info.GetRouteName(), info.FilterChainName(),
	)
	f.callbacks.Log(api.Info, msg)
}

// NullParser implements StreamFilterConfigParser. It is used when no configuration is provided.
type NullParser struct{}

func (NullParser) Parse(any *anypb.Any) (interface{}, error) {
	return nil, nil
}

func (NullParser) Merge(parentConfig interface{}, childConfig interface{}) interface{} {
	return nil
}

func CollectProfile() {
	for {
		time.Sleep(time.Minute)
		name := fmt.Sprintf("heap-%d.pprof.gz", time.Now().Unix())
		f, err := os.CreateTemp(os.TempDir(), name)
		if err != nil {
			continue
		}
		runtime.GC()
		_ = pprof.WriteHeapProfile(f)
		_ = f.Close()
	}
}
```

In provided archive [gofilter.zip](https://github.com/user-attachments/files/17575528/gofilter.zip) there is a docker-compose file with envoy with go filter, and simple go application that handles requests from envoy and sends request to envoy to load it.  To run it just run command `make docker-up`, after building images you should see something similar to
```
~> docker ps
CONTAINER ID   IMAGE            COMMAND                  CREATED          STATUS          PORTS       NAMES
f5bf730d777e   gofilter-envoy   "/docker-entrypoint.…"   56 seconds ago   Up 55 seconds   10000/tcp   gofilter-envoy-1
ad3b85d0acb6   gofilter-app     "/app/app"               56 seconds ago   Up 55 seconds   8181/tcp    gofilter-app-1
```
You could copy all profiles and visualize them with these commands.
```
~> docker cp f5bf730d777e:/tmp ./
~> go tool pprof -web tmp/heap-1730303069.pprof.gz814346379
```
Also if you check envoy logs there is no logs that I added in filter's implementation OnDestroy method.
```
docker logs f5bf730d777e
```

To stop containers run `make docker-down`

**Comment by spacewander** [comment]
@eerakli 
Thanks for your info. Sorry, I am busy working these days. Maybe I could investigate it at this weekend.

**Comment by eerakli** [comment]
I checked if this problem is actual for other version, and I can't reproduce it on version v1.29! Looks like there is something wrong with onDestroy in 1.27.

**Comment by spacewander** [comment]
I am glad to hear that the problem was fixed in 1.29.
I compare the golang filter between 1.27 and 1.29, the only change on envoyGoFilterOnHttpDestroy is https://github.com/envoyproxy/envoy/commit/0a2d0f7e89ca53c30fd42617c667160c62cfa400, which is memory leak fix merged on 1.28.

Would you try 1.28 to confirm if this problem is fixed by that change?
